### PR TITLE
ambient: allow cross-namespace waypoint references

### PR DESCRIPTION
**Pull request by howardjohn**
Our old strategy was to allow a user to set the label to
"namespace/name". This is great... but its illegal Kubernetes syntax,
making it impossible for a user to actually do this.

Instead, split this into two annotations

### CODE REVIEW

File: pkg/config/constants/constants.go
Comment (bleggett) [inline pkg/config/constants/constants.go:193]:
```suggestion
	// This allows cross-namespace waypoint references. If unset, the same namespace is assumed. If set, and AmbientUseWaypointLabel is unset, ignored.
```

**Review by bleggett — APPROVED** [review]
The simplest and most unsurprising way to do it, and so probably the best.

File: pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go
Comment (bleggett) [inline pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go:155]: Do we have unit tests around `getUseWaypoint`? We probably should.

**Comment by hzxuzhonghu** [comment]
Not sure in understand the motivation. The Original design was to not share waypoint globally.

**Comment by howardjohn** [comment]
@hzxuzhonghu the intent here is to give more users flexibility. While a default of 1-waypoint-per-namespace is a good starting point, it does not fit all use cases. For example, users may have very small namespaces (productpage namespace, reviews namespace, etc) and want to share a waypoint to reduce costs.

The most critical use case, IMO, is usage as an egress. For instance:

```
apiVersion: networking.istio.io/v1
kind: ServiceEntry
metadata:
  name: egress-example
  labels:
    istio.io/use-waypoint: egress
    istio.io/use-waypoint-namespace: shared-infra
spec:
  hosts:
  - httpbin.org
  location: MESH_INTERNAL
  ports:
  - name: http
    number: 80
    protocol: HTTP
  resolution: DNS
```

With this PR, this is enough to get all traffic to httpbin.org through the waypoint as an egress gateway, without the additional 4 configs typically required (Gateway, 2 virtual services, and DR)

**Review by costinm — COMMENTED** [review]
What happened with api/labels package ( which was supposed to generate docs and be reviewed as API ) ? 

Can we just move all labels under istio/istio ( since it's pretty clear we can't stop defining APIs in istio/istio ) - it was quite useful to have some way to know whatever we support without having to search in various 'constants' packages. 

Since 'cross namespace' is the topic - was there any discussion or review on the security implications of crossing this boundary ? Could this be used to send traffic to something that is not expecting it ( or doesn't want to handle the traffic ) ? In the K8S Gateway we have both the gateway and the route mutually agree, it was done for a reason - this is close and I'm worried we're missing something and will be hard to fix security later.

**Review by costinm — COMMENTED** [review]
Security review - and usability review. Is there any doc describing the use case, I can't even imagine why would someone want to allow users in random namespaces select waypoints in arbitrary namespaces - will be pretty hard to support and control. 

The use case of having a 'per cluster' waypoint - that the mesh owner defines in istio-system and is enforced, maybe with some label selectors to apply to specific namespaces and services - that's clear, but a different persona.

**Comment by costinm** [comment]
> @hzxuzhonghu the intent here is to give more users flexibility. While a default of 1-waypoint-per-namespace is a good starting point, it does not fit all use cases. For example, users may have very small namespaces (productpage namespace, reviews namespace, etc) and want to share a waypoint to reduce costs.
> 
> The most critical use case, IMO, is usage as an egress. For instance:
> 
> ```
> apiVersion: networking.istio.io/v1
> kind: ServiceEntry
> metadata:
>   name: egress-example
>   labels:
>     istio.io/use-waypoint: egress
>     istio.io/use-waypoint-namespace: shared-infra
> spec:
>   hosts:
>   - httpbin.org
>   location: MESH_INTERNAL
>   ports:
>   - name: http
>     number: 80
>     protocol: HTTP
>   resolution: DNS
> ```
> 
> With this PR, this is enough to get all traffic to httpbin.org through the waypoint as an egress gateway, without the additional 4 configs typically required (Gateway, 2 virtual services, and DR)

Using the labels for normal Service is one thing - using it for egress and ServiceEntry is a whole different level of risk,
 in particular given the current lack of controls around SE namespace isolation.

It would be major degradation on the already poor security posture we have around this. 

Not against the intended use case - but expressing it in random namespaces by the namespace owners is against 
one of the main reasons egress is used ( to enforce outbound policies ).

**Comment by bleggett** [comment]
> Can we just move all labels under istio/istio ( since it's pretty clear we can't stop defining APIs in istio/istio ) - it was quite useful to have some way to know whatever we support without having to search in various 'constants' packages.

For this, I would suggest we just make sure every new label goes with a public docs PR, at least (and crossref the PRs). We can do more, that should probably be the minimum.

**Comment by costinm** [comment]
On Wed, Jul 3, 2024 at 9:33 AM Ben Leggett ***@***.***> wrote:

> Can we just move all labels under istio/istio ( since it's pretty clear we
> can't stop defining APIs in istio/istio ) - it was quite useful to have
> some way to know whatever we support without having to search in various
> 'constants' packages.
>
> For this, I would suggest we just make sure every new label goes with a
> public docs PR, at least (and crossref the PRs). We can do more, that
> should probably be the minimum
>

That may be worse - because there is wide belief that public docs imply a
feature is intended for public usage ( and is supported, etc).

There is a middle ground between 'every label or feature is publicly
available and long term supported from start' and 'hide APIs in random
constant files with no API review ( UX, security, etc) and later
claim they're in use and can't fix anything'.

What I am asking for is a way for a Istio maintainer/developer to discover
all the APIs and to have some way to flag potential 'APIs' and do a deeper
review - as a first step.
Having some docs in istio/istio would be nice too - but certainly not in
istio.io until the feature is considered stable.

But I think for this particular PR ( and the original PR which may or may
not have discussed the API implications) - the bigger concern is
security/UX - not where it is documented or placed.

**Comment by bleggett** [comment]
> That may be worse - because there is wide belief that public docs imply a
feature is intended for public usage ( and is supported, etc).

yes, I am suggesting in general that we shouldn't add labels to the codebase until we are ready to do this.

> But I think for this particular PR ( and the original PR which may or may not have discussed the API implications) - the bigger concern is security/UX - not where it is documented or placed.

:+1:

**Comment by howardjohn** [comment]
@costinm probably worth broader discussion, but at a high level do you think it makes sense to only allow if it is unified with allowedRoutes setting in the route?

**Comment by costinm** [comment]
Not necessarily allowedRoutes - it's a different use case.

But any equivalent (maybe simpler) form of mutual consent, with the default as isolated.

My main proposal (in context of DNS and autoalloc) is to have at least a setting that restrict SE with arbitrary FQDN and frontend role to one namespace, with clear labeling and mesh owner setting. Not aware of any other way to secure dns interception and SE in general. 

There is no harm in allowing the user to label things and use RBAC and namespaces.

**Comment by costinm** [comment]
Sorry - jumped to SE without the middle step - one way to handle the cross namespace waypoint and egress is to use SE as a way to indicate the external name. 

You may remember the discussion on using Gateway objects and the FQDN ( which can be the service name ) - I no longer think it's a good idea, labeled SE are better and we already use them for egress. They just need hardening.

**Comment by howardjohn** [comment]
> Not necessarily allowedRoutes - it's a different use case.
> 
> But any equivalent (maybe simpler) form of mutual consent, with the default as isolated.
> 
> My main proposal (in context of DNS and autoalloc) is to have at least a setting that restrict SE with arbitrary FQDN and frontend role to one namespace, with clear labeling and mesh owner setting. Not aware of any other way to secure dns interception and SE in general.
> 
> There is no harm in allowing the user to label things and use RBAC and namespaces.

I feel like the FQDN and SE stuff is orthogonal. Everything in this PR applies to Service as well for instance. 

Allow routes actually doesn't seem too bad to me - effectively the services are routes, and it gives the exact level of control we need in the right API

**Comment by costinm** [comment]
FQDN and SE are similar because it has the same 'security' problem ( or
worse ), breaking
isolation and ownership.

I think (like unfortunately many other things in Istio) the level of
control is far too broad and lax,
providing too much power and too little security and operational controls.

We start with a reasonable use case - 'per cluster' or 'shared' Waypoints.
This can be safely expressed
by the mesh admin ( who is really the one that should decide ) with
something in MeshConfig or
an Istiod setting. Admin can also define label selectors ( like we use for
injection ) so only certain
namespaces or services use the shared waypoint. No change in the helm
charts or manifests for the deployed apps,
majority of the users are not impacted.

Do we have a use case for 'any namespace owner should be able to managed
waypoints and any
service owner must pick arbitrary namespaces that will get configs for
their workloads' ? Maybe, but
we can start by implementing the safe, simple, clean - and more common -
use case with a simple solution,
and have a very serious security review on the risks for the second use
case ( if we actually find users who
provide feedback that the first is not sufficient ).

There is nothing wrong with users adding labels to any CR - to convey
intent, readability, integrate
with OPA and other tools that use label selectors. Not ALL Services and
Namespaces are the
same - not all Services are routes or frontends or both TCP and HTTP.   But
everything must be
mutual - users in a namespace may express their intent with labels, but the
other side ( Waypoint
 owner, domain owner,  etc)  in a different namespace must also express
their intent to allow by
using a label selector. Only exception should be istio-system ( == mesh
admin ), which may
unilaterally set policies.

The problem with SE ( and Service) is that it lacks clarity on user intent
- is it intended as a
backend, is it intended as for HTTP/L7, is it global-capable,
regional/local ?

So back to this PR: can we start with implementing the narrow use case
first ?

**Comment by howardjohn** [comment]
I 100% agree we need a mutual handshake. I didn't think much about it initially on this PR since it was just fixing a bug that the already-implemented cross-ns support broke at some point, but definitely should improve it before we fix it. 

What is less clear to me is why not use allowed routes. If we look at "waypoint the hard way", that is exactly how it would work - the gateway owner allows a namespace to bind to it, and the namespace owner chooses to expose something on the gateway. The same would apply here. I am not sure why that should require mesh admin permission? 

I don't think the use cases are only a *single* waypoint for the entire cluster (which, realistically, is probably only a good idea for cloud waypoints) or a single egress gateway. Waypoint per app (where app spans a few namespace), etc are pretty viable cost saving mechanisms. In large scale deployments even egress gateway is not a singleton but many smaller ones scoped in various ways. I worry if we require mesh admin config to support these it's more complex, less flexible, and less standard (differs from how normal Gateway do the handshake).

Just to be clear, what I am suggesting is we do what is in this PR but add an additional check that the Gateway allowedRoutes must permit the binding. The default allowedRoutes is " same namespace ", do a waypoint owner must opt in to sharing

**Comment by costinm** [comment]
On Sat, Jul 6, 2024, 09:55 John Howard ***@***.***> wrote:

> I 100% agree we need a mutual handshake. I didn't think much about it
> initially on this PR since it was just fixing a bug that the
> already-implemented cross-ns support broke at some point, but definitely
> should improve it before we fix it.
>
> What is less clear to me is why not use allowed routes. If we look at
> "waypoint the hard way", that is exactly how it would work - the gateway
> owner allows a namespace to bind to it, and the namespace owner chooses to
> expose something on the gateway. The same would apply here. I am not sure
> why that should require mesh admin permission?
>

I meant: for the simple/common case, all we need is mesh admin to indicate
a namespace for the 'cluster waypoint' ( which can also be used for cluster
egress). He may as well indicate that all http services will use it.

For the most flexible case, where a random namespace admin deploys a
waypoint and indicate a set of namespaces allowed to use it - yes,
allow-routes is the right API and works. Just a bigger pain - I bet
majority of users will just do an allow all.

The problem with current model of defining an istio specific label is that
it requires user to modify helm charts or use complex values.yaml. Not easy
at all.

There is an argument that using a separate CR ( like a real http route ) is
harder than just setting a label - but that's true for demos and examples,
when you deal with real helm charts good luck finding their way of adding
labels to services ( if they support it ) or doing kustomize or other
stuff. Not all k8s helm charts are written with Istio in mind.

Having waypoint in arbitrary namespace, with allow routes and httproute
objects is standard Gateway and works pretty well without requiring any
change to the helm charts. But still a lot of work for a made-up use case,
when a per cluster waypoint and egress is so much common and easy.








I don't think the use cases are only a *single* waypoint for the entire

**Comment by howardjohn** [comment]
I've implemented selection based on allowedRoutes.

One issue with this is it is hard to debug. https://github.com/istio/istio/pull/51945 fixes that.

**Review by stevenctl — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
>I meant: for the simple/common case, all we need is mesh admin to indicate
a namespace for the 'cluster waypoint' ( which can also be used for cluster
egress). He may as well indicate that all http services will use it.

>For the most flexible case, where a random namespace admin deploys a
waypoint and indicate a set of namespaces allowed to use it - yes,
allow-routes is the right API and works. Just a bigger pain - I bet
majority of users will just do an allow all.

In terms of UX, bi-directional setting to use a waypoint or any infra is a pain.  Generally a separate namespace should be enough for a small company bussiness.  More fine-grained control should be with `ReferenceGrant`

These labels "istio.io/use-waypoint-namespace" and "istio.io/waypoint-for" may be not that safe.

**Comment by hzxuzhonghu** [comment]
@howardjohn Why use Gateway's AlloweRoutes to determine whether a service/pod can use the waypoint. I think it is not the intended semantic of API

**Comment by LuyaoZhong** [comment]
@howardjohn the waypoint seems not support traffic management across namespaces, when I apply some virtual service and destination rule config, waypoint didn't get correct xDS response from istiod, I'd like to know if this is a bug or a wrong usage.

**Comment by costinm** [comment]
BTW - see my doc on ServiceEntry security, for the httpbin.com example in this PR it should be MESH_EXTERNAL and 
should be used differently for 'backend' role ( no need of Waypoint and safe for users to add to untrusted namespace) and fronted / interception - where DNS interception and mesh-wide impact are breaking isolation. Also not clear if 'egress control' should be controlled by untrusted users labeling (or not) Services or by the mesh/egress admin.

The PR is fine for Service or real MESH_INTERNAL ServiceEntries ( with the restrictions suggested in the doc to preserve boundaries ).

We should also make it clear that using this mechanism requires the use of HttpRoute and DR in the waypoint namespace, with gateway-style attachment. DR in the Service namespace can't be attached to the (implicit or explicit) route in the gateway/waypoint namespace.

**Comment by howardjohn** [comment]
> @howardjohn the waypoint seems not support traffic management across namespaces, when I apply some virtual service and destination rule config, waypoint didn't get correct xDS response from istiod, I'd like to know if this is a bug or a wrong usage.

Which namespace is the config in, and what is it applying to? Did you try with HTTPRoute

**Comment by LuyaoZhong** [comment]
> > @howardjohn the waypoint seems not support traffic management across namespaces, when I apply some virtual service and destination rule config, waypoint didn't get correct xDS response from istiod, I'd like to know if this is a bug or a wrong usage.
> 
> Which namespace is the config in, and what is it applying to? Did you try with HTTPRoute

The config was in the service namespace, and the waypoint was in a diff namespace, and HTTPRoute was used. 
I got that currently there is some limitation as @costinm mentioned, so I refined my env but it still doesn't work.

1. bookinfo services in default namespace
 `kubectl apply -f samples/bookinfo/platform/kube/bookinfo.yaml -n default`
 `kubectl apply -f samples/bookinfo/platform/kube/bookinfo-versions.yaml -n default`
2. waypoint gateway in istio-system namespace 
`kubectl apply -f shared-waypoint.yaml -nistio-system`
```shell
# cat shared-waypoint.yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: waypoint
  namespace: istio-system
spec:
  gatewayClassName: istio-waypoint
  listeners:
  - name: mesh
    port: 15008
    protocol: HBONE
    allowedRoutes:
      namespaces:
        from: All
``` 

3. httproute config in istio-system namespace
`kubectl apply -f route-reviews-v1.yaml -nistio-system`
```shell
# cat route-reviews-v1.yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: reviews
spec:
  parentRefs:
  - group: ""
    kind: Service
    name: reviews.default.svc.cluster.local
    port: 9080
  rules:
  - backendRefs:
    - name: reviews-v1.default.svc.cluster.local
      port: 9080
---
```

4. label default namespace to use waypoint in istio-system namespace
```shell
# kubectl get ns default --show-labels
NAME      STATUS   AGE   LABELS
default   Active   61d   istio.io/dataplane-mode=ambient,istio.io/use-waypoint-namespace=istio-system,istio.io/use-waypoint=waypoint,kubernetes.io/metadata.name=default
```
5. expect all traffic routed to reviews-v1 but failed

```shell
# kubectl exec deploy/sleep   -- sh -c "for i in \$(seq 1 10); do curl -s http://productpage.default.svc.cluster.local:9080/productpage; done" |grep reviews
        <u>reviews-v1-5fd6d4f8f8-q4mxd</u>
        <u>reviews-v3-7d99fd7978-5rpjh</u>
        <u>reviews-v3-7d99fd7978-5rpjh</u>
        <u>reviews-v1-5fd6d4f8f8-q4mxd</u>
        <u>reviews-v1-5fd6d4f8f8-q4mxd</u>
        <u>reviews-v3-7d99fd7978-5rpjh</u>
        <u>reviews-v1-5fd6d4f8f8-q4mxd</u>
        <u>reviews-v2-6f9b55c5db-jxsz2</u>
        <u>reviews-v3-7d99fd7978-5rpjh</u>
        <u>reviews-v1-5fd6d4f8f8-q4mxd</u>
```


@howardjohn could you help figure out which step is wrong? 
Besides, I saw this cross-namespace waypoint feature in Istio 1.23 anouncement, but no docs about it. This feature is a big change, it will be great if we can have documentation about the feature, including limitations, use cases and the examples.

**Comment by LuyaoZhong** [comment]
httproute config is not correct in step 3, the traffic goes correct way after fixing it.

``` yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: reviews
spec:
  parentRefs:
  - group: ""
    kind: Service
    name: reviews
    namespace: default
    port: 9080
  rules:
  - backendRefs:
    - name: reviews-v1
      namespace: default
      port: 9080
---
```

**Pull request merged**
